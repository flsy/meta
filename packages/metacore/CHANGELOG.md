# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.0.0-alpha.6](https://github.com/flsy/meta/compare/@falsy/metacore@1.0.0-alpha.5...@falsy/metacore@1.0.0-alpha.6) (2022-08-15)

**Note:** Version bump only for package @falsy/metacore





# [1.0.0-alpha.5](https://github.com/flsy/meta/compare/@falsy/metacore@1.0.0-alpha.4...@falsy/metacore@1.0.0-alpha.5) (2022-08-15)

**Note:** Version bump only for package @falsy/metacore





# [1.0.0-alpha.4](https://github.com/flsy/meta/compare/@falsy/metacore@1.0.0-alpha.3...@falsy/metacore@1.0.0-alpha.4) (2021-11-24)

**Note:** Version bump only for package @falsy/metacore





# [1.0.0-alpha.3](https://github.com/flsy/meta/compare/@falsy/metacore@1.0.0-alpha.2...@falsy/metacore@1.0.0-alpha.3) (2021-11-24)

**Note:** Version bump only for package @falsy/metacore





# [1.0.0-alpha.2](https://github.com/flsy/meta/compare/@falsy/metacore@1.0.0-alpha.1...@falsy/metacore@1.0.0-alpha.2) (2021-10-29)

**Note:** Version bump only for package @falsy/metacore





# [1.0.0-alpha.1](https://github.com/flsy/meta/compare/@falsy/metacore@1.0.0-alpha.0...@falsy/metacore@1.0.0-alpha.1) (2021-10-19)

**Note:** Version bump only for package @falsy/metacore





# [1.0.0-alpha.0](https://github.com/flsy/meta/compare/@falsy/metacore@1.0.0...@falsy/metacore@1.0.0-alpha.0) (2021-10-19)

**Note:** Version bump only for package @falsy/metacore





## [0.0.10](https://github.com/flsy/meta/compare/@falsy/metacore@0.0.9...@falsy/metacore@0.0.10) (2021-09-22)

**Note:** Version bump only for package @falsy/metacore





## [0.0.9](https://github.com/flsy/meta/compare/@falsy/metacore@0.0.8...@falsy/metacore@0.0.9) (2021-09-22)

**Note:** Version bump only for package @falsy/metacore





## [0.0.8](https://github.com/flsy/meta/compare/@falsy/metacore@0.0.6...@falsy/metacore@0.0.8) (2021-09-22)

**Note:** Version bump only for package @falsy/metacore





## [0.0.7](https://github.com/flsy/meta/compare/@falsy/metacore@0.0.6...@falsy/metacore@0.0.7) (2021-07-28)

**Note:** Version bump only for package @falsy/metacore





## [0.0.6](https://github.com/flsy/meta/compare/@falsy/metacore@0.0.5...@falsy/metacore@0.0.6) (2021-07-12)

**Note:** Version bump only for package @falsy/metacore





## [0.0.5](https://github.com/flsy/meta/compare/@falsy/metacore@0.0.4...@falsy/metacore@0.0.5) (2021-07-08)

**Note:** Version bump only for package @falsy/metacore





## [0.0.4](https://github.com/flsy/meta/compare/@falsy/metacore@0.0.3...@falsy/metacore@0.0.4) (2021-07-08)

**Note:** Version bump only for package @falsy/metacore





## [0.0.3](https://github.com/flsy/meta/compare/@falsy/metacore@0.0.2...@falsy/metacore@0.0.3) (2021-03-16)

**Note:** Version bump only for package @falsy/metacore





## 0.0.2 (2021-03-12)

**Note:** Version bump only for package @falsy/metacore
