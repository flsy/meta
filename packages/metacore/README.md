# `metacore`

## Usage

```
const metacore = require('metacore');

// TODO: DEMONSTRATE API
```
