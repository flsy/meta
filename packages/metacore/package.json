{
  "name": "@falsy/metacore",
  "version": "1.0.0-alpha.4",
  "description": "> TODO: description",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "build": "tsc",
    "lint": "echo todo",
    "test": "echo todo"
  },
  "author": "Honza Smid <jansmidhonza@gmail.com>",
  "license": "ISC",
  "homepage": "https://github.com/flsy/meta/tree/main/packages/metacore#readme",
  "directories": {
    "lib": "lib",
    "test": "__tests__"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/flsy/meta.git"
  },
  "bugs": {
    "url": "https://github.com/flsy/meta/issues"
  },
  "devDependencies": {
    "typescript": "~4.5.3"
  }
}
