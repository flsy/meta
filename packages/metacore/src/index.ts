export * from './interfaces';
export * from './metatable/interfaces';
export * from './metafilters/interfaces';
export * from './metaforms/interfaces';
export * from './metaforms/validation';
