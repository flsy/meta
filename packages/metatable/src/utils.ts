import { Column, Columns, OneOrMany } from './interfaces';
import {
  defaultTo, has,
  head,
  isNil,
  lensPath,
  mergeRight,
  not,
  pipe,
  prop,
  set,
  view,
  when,
} from 'ramda';
import { getFormData } from 'metaforms';

export const getCellValue = <TRow>(bits: string[]) => (object: TRow): OneOrMany<any> => {
  const [property, ...rest] = bits;

  if(!property) {
    return object;
  }

  const value = prop<string, any>(property)(object);
  if (Array.isArray(value)) {
    return value.map((v) => getCellValue(rest)(v));
  }
  if (value && head(rest)) {
    return getCellValue(rest)(value);
  }
  return value;
};

export const isColumn = <TTypes>(column: Column<TTypes> | Columns<TTypes>): column is Column<TTypes> => typeof column.type === 'string';
export const getColumnPaths = <TColumns extends Columns<TTypes>, TTypes>(columns: TColumns, parentKey: string[] = []): string[][] =>
  Object.entries(columns).reduce((acc, [name, column]) => {
    if (isColumn(column)) {
      return [...acc, [...parentKey, name]];
    }

    return [...acc, ...getColumnPaths(column, [...parentKey, name])];
  }, []);

export const findColumnPath = <TColumns extends Columns<TTypes>, TTypes>(find: (column: Column<TTypes>) => boolean) => (columns: TColumns): string[] =>
  defaultTo([], getColumnPaths(columns).find((colPath) => find(view(lensPath(colPath))(columns))));

export const filterColumnPaths = <TColumns extends Columns<TTypes>, TTypes>(filter: (column: Column<TTypes>) => boolean) => (columns: TColumns): string[][] =>
  defaultTo([], getColumnPaths(columns).filter((colPath) => filter(view(lensPath(colPath), columns))));


// eslint-disable-next-line @typescript-eslint/ban-types
const setSortFormValue = (sortForm: object): object => {
  return Object.entries(sortForm).reduce((acc, [key, field]) => {
    if(has('type', field) && prop('type', field) === 'sort') {
      return { ...acc, [key]: { ...field, value: undefined } };
    }

    if(field && typeof field === 'object') {
      return { ...acc, [key]: { ...field, ...setSortFormValue(field) } };
    }

    return acc;
  }, {})
}

export const unsetAllSortFormValues = <TColumns extends Columns<TTypes>, TTypes>(columns: TColumns): TColumns => {
  return getColumnPaths(columns).reduce((acc, colPath) => {
    const columnSortFormLens = lensPath([...colPath, 'sortForm']);
    const columnSortForm = view(columnSortFormLens, columns);

    if(columnSortForm) {
      // eslint-disable-next-line @typescript-eslint/ban-types
      return set(columnSortFormLens, setSortFormValue(columnSortForm as object), acc);
    }

    return acc;
  }, columns)
}

// eslint-disable-next-line @typescript-eslint/ban-types
export const toMetaFilters = <TColumns extends Columns<TTypes>, TTypes>(columns: TColumns): { filters: object; sort: object } => {
  return getColumnPaths(columns).reduce((acc, columnPath) => {
    const columnFilterForm = view(lensPath([...columnPath, 'filterForm']), columns);
    const columnSortForm = view(lensPath([...columnPath, 'sortForm']), columns);

    const filterFormData = when(
      pipe(isNil, not),
      getFormData
    )(columnFilterForm);

    const sortFormData = when(
      pipe(isNil, not),
      getFormData
    )(columnSortForm);

    const filters = mergeRight(acc.filters, filterFormData);
    const sort = mergeRight(acc.sort, sortFormData);

    return { filters, sort };
  }, { filters: {}, sort: {} });
};


// eslint-disable-next-line @typescript-eslint/ban-types
const setNestedForm = (path: string[], form: object) => {
  const pathWFields = path.reduce((acc, p) => [...acc, p, 'fields'], [])

  const o = pathWFields.reduce((result, p, i) => {
    if(p === 'fields') {
      const currPath = pathWFields.slice(0, i);
      return set(lensPath(currPath), { type: 'group' }, result);
    }

    return result;
  }, {})


  return set(lensPath(pathWFields), form, o);
}


type StringFilterValue = Array<{ operator: 'EQ' | 'LIKE', value: string }>;
export const getStringFilter = (path: string[], value: StringFilterValue) =>
  setNestedForm(path, {
    type: {
      type: 'hidden',
      value: 'string',
    },
    filters: {
      type: 'string-filter',
      value
    },
    submit: {
      type: 'submit',
      label: 'submit',
    },
  })