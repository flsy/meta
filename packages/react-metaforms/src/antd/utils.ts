import {
    TextMetaProps,
    NumberMetaProps,
    PasswordMetaProps,
    SelectMetaProps,
    TextareaMetaProps,
    CheckboxMetaProps,
    SubmitMetaProps,
    DateMetaProps,
    ImageMetaProps,
    FileMetaProps,
    DateRangeMetaProps,
    MultiSelectMetaProps,
    AutocompleteMetaProps,
    MetaField,
    ObjectMetaProps,
    HiddenMetaProps,
    ActionMetaProps,
    IThreeStateSwitch, LayoutMetaProps,
} from '@falsy/metacore';

export const isText = (field: MetaField): field is TextMetaProps => field.type === 'text';
export const isNumber = (field: MetaField): field is NumberMetaProps => field.type === 'number';
export const isHidden = (field: MetaField): field is HiddenMetaProps => field.type === 'hidden';
export const isPassword = (field: MetaField): field is PasswordMetaProps => field.type === 'password';
export const isSelect = (field: MetaField): field is SelectMetaProps => field.type === 'select';
export const isTextarea = (field: MetaField): field is TextareaMetaProps => field.type === 'textarea';
export const isCheckbox = (field: MetaField): field is CheckboxMetaProps => field.type === 'checkbox';
export const isSubmit = (field: MetaField): field is SubmitMetaProps => field.type === 'submit';
export const isDate = (field: MetaField): field is DateMetaProps => field.type === 'date';
export const isImage = (field: MetaField): field is ImageMetaProps => field.type === 'image';
export const isFile = (field: MetaField): field is FileMetaProps => field.type === 'file';
export const isDateRange = (field: MetaField): field is DateRangeMetaProps => field.type === 'dateRange';
export const isMultiselect = (field: MetaField): field is MultiSelectMetaProps => field.type === 'multiselect';
export const isAutocomplete = (field: MetaField): field is AutocompleteMetaProps => field.type === 'autocomplete';
export const isObject = (field: MetaField): field is ObjectMetaProps => field.type === 'object';
export const isAction = (field: MetaField): field is ActionMetaProps => field.type === 'action';
export const isLayout = (field: MetaField): field is LayoutMetaProps => field.type === 'layout';
export const isThreeStateSwitch = (field: MetaField): field is IThreeStateSwitch => field.type === 'threeStateSwitch';
