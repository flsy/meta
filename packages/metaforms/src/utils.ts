import { find, isRight, Left, Maybe, propEq, Right } from 'fputils';
import { Validation } from '@falsy/metacore';
import { MetaFormErrorMessages, MetaField } from '@falsy/metacore';
import { lensPath, set, view, Lens, curry } from 'ramda';

export const isRequired = (validationRules: Validation[] = []): boolean => !!find(propEq('type', 'required'), validationRules);
export const hasError = (fields: MetaField[]): boolean => Object.keys(getErrorMessages(fields)).length > 0;

export const fieldPropertyLens = curry((property: string, fieldName: string, fields: MetaField[]): Maybe<Lens<any, any>> => {
  try {
    const index = fields.findIndex((field) => field.name === fieldName);

    if(index >= 0) {
      return Right(lensPath([index, property]));
    }

    return Left(new Error('Field not found'));
  } catch (e) {
    return Left(e as Error);
  }
});

export const setFieldProperty = curry((property: string, fieldName: string, value: any, fields: MetaField[]): MetaField[] => {
  const optionsLens = fieldPropertyLens(property);
  const lens = optionsLens(fieldName, fields);
  if(isRight(lens)) {
    return set(lens.value, value, fields);
  }

  return fields;
});

export const getFieldProperty = curry((property: string, fieldName: string, fields: MetaField[]): any => {
  const optionsLens = fieldPropertyLens(property);
  const lens = optionsLens(fieldName, fields);
  if(isRight(lens)) {
    return view(lens.value, fields);
  }

  return;
});

export const setFieldValidation = curry((fieldName: string, value: Validation[], fields: MetaField[]): MetaField[] =>
  setFieldProperty('validation', fieldName, value, fields)
);

export const getErrorMessages = (fields: MetaField[]): MetaFormErrorMessages =>
  fields.reduce((acc, f) => {
    const nameLens = lensPath(f.name.split('.'));
    return (f as any).errorMessage ? set(nameLens, (f as any).errorMessage, acc) : acc;
  }, {});
