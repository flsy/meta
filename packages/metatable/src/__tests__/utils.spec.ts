 import { filterColumnPaths, getStringFilter, unsetAllSortFormValues, toMetaFilters } from '../utils';
import { Columns, MetaField } from '@falsy/metacore';

const customerIdFilterForm: MetaField[] = [
  {
    name: 'customer.id.type',
    type: 'hidden',
    value: 'string'
  },
  {
    name: "customer.id.filters",
    type: 'text',
    value: [{ operator: 'EQ', value: 'CUST_ID' }],
  },
  {
    name: "customer.id.actions",
    type: "buttonGroup",
    items: [{
      name: 'submit',
      type: 'submit',
      label: 'Submit',
    }]
  },
];

const mockedColumns1: Columns<'number' | 'timestamp' | 'string' | 'boolean'> = {
  id: {
    type: 'number',
    label: 'Id',
    key: true,
    sortForm: [{
        name: 'id',
        type: 'sort',
      },
    ],
  },
  createdAtFormatted: {
    type: 'timestamp',
    label: 'CreatedAt',
    sortForm: [
      {
        name: 'createdAt',
        type: 'sort',
      },
    ]
  },
  createdBy: {
    name: {
      type: 'string',
      label: 'CreatedBy',
      sortForm: [{
        name: "createdBy.name",
        type: 'sort',
        value: 'ASC',
      }],
        filterForm: getStringFilter(['name'], {
          value: [{ value: 'hey', operator: 'EQ' }],
          label: 'Name',
          submit: {
            name: "name.actions",
            type: "buttonGroup",
            items: [
              {
                label: "Reset",
                name: "reset",
                type: "reset",
              },
              {
                name: 'submit',
                type: 'submit',
                label: 'Submit',
                primary: true
              },
            ]
          }}),
    },
  },
  attachments: {
    content: {
      type: 'string',
      label: 'Attachment content',
    },
    description: {
      type: 'string',
      label: 'Attachment description',
    },
  },
  isPublished: {
    type: 'boolean',
    label: 'Is published',
  },
  content: {
    type: 'string',
    label: 'Content',
    sortForm: [{
      name: 'content',
      type: 'sort',
      value: 'ASC',
    }],
    filterForm: [{
        name: 'content',
        label: 'Kontent',
        type: 'text',
        value: null,
        placeholder: 'Obsah',
        errorMessage: null,
      },
      {
        name: 'submit',
        type: 'submit',
        label: 'Ulo≈æit',
      },
    ],
  },
};
const mockedColumns2: Columns<'number' | 'string' | 'boolean'> = {
  id: {
    label: 'id',
    type: 'number',
    key: true,
    isOmitted: true,
    sortForm: [{
        name: 'id',
        type: 'sort',
        value: 'ASC',
      },
    ],
  },
  name: {
    label: 'name',
    type: 'string',
    key: false,
    sortForm: [{
        name: 'id',
        type: 'sort',
        value: 'ASC',
      },
    ],
  },
  customer: {
    id: {
      label: 'customer id',
      type: 'string',
      key: false,
      filterForm: customerIdFilterForm,
      sortForm: [{
          name: 'name',
          type: 'sort',
        },
      ],
    },
  },
  isDeleted: {
    label: 'is deleted',
    type: 'boolean',
    key: false,
    isOmitted: true,
    filterForm: [
      {
        name: 'isDeleted.type',
        type: 'hidden',
        value: 'boolean',
      },
      {
        name: 'isDeleted.value',
        type: 'checkbox',
        value: false,
      },
      {
        name: 'submit',
        type: 'submit',
        label: 'submit',
      }
    ]
  },
};

const col2toMetafilterOutput = {
  sort: {
    id: 'ASC',
  },
  filters: {
    isDeleted: {
      type: 'boolean',
      value: false
    },
    customer: {
      id: {
        type: 'string',
        filters: [{ operator: 'EQ', value: 'CUST_ID' }],
      },
    },
  },
};

describe('Metatable utils', () => {
  it('transform to meta filters', async () => {
    expect(toMetaFilters(mockedColumns2)).toMatchObject(col2toMetafilterOutput);
  });

  it('gets string filter without value, label and submit', () => {
    expect(getStringFilter(['a', 'b'])).toEqual([
      {
        name: "a.b.type",
        type: "hidden",
        value: "string"
      },
      {
        name: "a.b.filters",
        type: "text"
      }
    ])
  })

  it('gets string filter', async () => {
    expect(getStringFilter(['customer', 'id'], {
      value: [{ operator: 'EQ', value: 'CUST_ID' }],
      label: 'Customer id',
      submit: {
        name: "customer.id.actions",
        type: "buttonGroup",
        items: [{
          name: 'submit',
          type: 'submit',
          label: 'Submit',
        }]
      }}
    )).toMatchObject(customerIdFilterForm);
  });

  it('filters column path', async () => {
    const displayedColumnPaths = filterColumnPaths((column) => !column?.isOmitted)(mockedColumns2);
    expect(displayedColumnPaths).toEqual([['name'], ['customer', 'id']]);
  });
  it('unsetAllSortFormValues simple columns', async () => {
    const result2 = unsetAllSortFormValues(mockedColumns2);

    expect(result2.id).toEqual({
      isOmitted: true,
      key: true,
      label: 'id',
      sortForm: [
        {
          name: 'id',
          type: 'sort',
          value: undefined,
        },
      ],
      type: 'number',
    });

    expect(result2.name).toEqual({
      key: false,
      label: 'name',
      sortForm: [
        {
          name: 'id',
          type: 'sort',
          value: undefined,
        },
      ],
      type: 'string',
    });

    expect(result2.customer['id']).toEqual({
      key: false,
      label: 'customer id',
      sortForm: [
        {
          name: 'name',
          type: 'sort',
          value: undefined,
        },
      ],
      type: 'string',
      filterForm: [
        {
          name: "customer.id.type",
          type: "hidden",
          value: "string"
        },
        {
          name: "customer.id.filters",
          type: "text",
          value: [
            {
              operator: "EQ",
              value: "CUST_ID"
            }
          ]
        },
        {
          items: [
            {
              label: "Submit",
              name: "submit",
              type: "submit"
            }
          ],
          name: "customer.id.actions",
          type: "buttonGroup"
        }
      ],
    });

    expect(result2.isDeleted).toEqual({
      filterForm: [
        {
          name: "isDeleted.type",
          type: "hidden",
          value: "boolean"
        },
        {
          name: "isDeleted.value",
          type: "checkbox",
          value: false
        },
        {
          label: "submit",
          name: "submit",
          type: "submit"
        }
      ],
      isOmitted: true,
      key: false,
      label: 'is deleted',
      type: 'boolean',
    });
  });
  it('unsetAllSortFormValues nested columns', async () => {
    const result2 = unsetAllSortFormValues(mockedColumns1);

    expect(result2.id).toEqual({
      key: true,
      label: 'Id',
      sortForm: [
        {
          name: 'id',
          type: 'sort',
          value: undefined,
        },
      ],
      type: 'number',
    });

    expect(result2.createdBy).toEqual({
      name: {
        filterForm: [
          {
            name: "name.type",
            type: "hidden",
            value: "string"
          },
          {
            label: 'Name',
            name: "name.filters",
            type: "text",
            value: [
              {
                operator: "EQ",
                value: "hey"
              }
            ]
          },
          {
            items: [
              {
                label: "Reset",
                name: "reset",
                type: "reset",
              },
              {
                label: "Submit",
                name: "submit",
                primary: true,
                type: "submit"
              }
            ],
            name: "name.actions",
            type: "buttonGroup"
          }
        ],
        label: 'CreatedBy',
        sortForm: [{
            name: 'createdBy.name',
            type: 'sort',
            value: undefined,
          },
        ],
        type: 'string',
      },
    });
  });
});
