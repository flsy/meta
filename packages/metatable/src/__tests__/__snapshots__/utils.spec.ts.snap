// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metatable utils sets sort form value 2 1`] = `
Object {
  "attachments": Object {
    "content": Object {
      "label": "Attachment content",
      "type": "string",
      "value": undefined,
    },
    "description": Object {
      "label": "Attachment description",
      "type": "string",
      "value": undefined,
    },
  },
  "content": Object {
    "filterForm": Object {
      "content": Object {
        "errorMessage": null,
        "label": "Kontent",
        "placeholder": "Obsah",
        "type": "text",
        "value": null,
      },
      "submit": Object {
        "label": "Uložit",
        "type": "submit",
        "value": undefined,
      },
    },
    "label": "Content",
    "sortForm": Object {
      "content": Object {
        "type": "sort",
        "value": "ASC",
      },
    },
    "type": "string",
  },
  "createdAtFormatted": Object {
    "label": "CreatedAt",
    "sortForm": Object {
      "createdAt": Object {
        "type": "sort",
        "value": undefined,
      },
    },
    "type": "timestamp",
  },
  "createdBy": Object {
    "name": Object {
      "filterForm": Object {
        "name": Object {
          "errorMessage": null,
          "label": "Name",
          "placeholder": "name placeholder",
          "type": "text",
          "value": null,
        },
        "submit": Object {
          "label": "Uložit",
          "type": "submit",
          "value": undefined,
        },
      },
      "label": "CreatedBy",
      "sortForm": Object {
        "name": Object {
          "type": "sort",
        },
      },
      "type": "string",
    },
  },
  "id": Object {
    "key": true,
    "label": "Id",
    "sortForm": Object {
      "id": Object {
        "type": "sort",
        "value": "DESC",
      },
    },
    "type": "number",
  },
  "isPublished": Object {
    "label": "Is published",
    "type": "boolean",
    "value": undefined,
  },
}
`;
